### PR TITLE
Add tests using checkpoint restarts and multiple sub apps

### LINKED ISSUE #15287: Restart multiple instantiations of a MultiApp from checkpoint
## Reason

A typical multiphysics setup is to first obtain a multiapp steady-state solution and to restart from that solution as the initial condition for a transient calculation. Checkpoint files are often used to restart.

If the master app only instantiates a given sub-app once, then the sub-apps can be restarted as follows (see [here](https://github.com/idaholab/moose/tree/next/test/tests/restart/restart_transient_from_steady) for more details):

```
[Mesh]
  file = master_out_subapp0_checkpoint_cp/LATEST
[]


[Problem]
  restart_file_base = master_out_subapp0_checkpoint_cp/LATEST
  skip_additional_restart_data = true
  force_restart = true
[]
``` 

However, if that sub-app is instantiated in multiple locations, then multiple `master_out_subapp*_checkpoint_cp` checkpoint folders will be created and it is not currently possible to have a single transient sub-app input file that will pick up the correct folder based on location.

## Design

Ideally, it would be great to have something like this in the transient sub-app input file:

```
[Mesh]
  file = master_out_subapp*_checkpoint_cp/LATEST
[]


[Problem]
  restart_file_base = master_out_subapp*_checkpoint_cp/LATEST
  skip_additional_restart_data = true
  force_restart = true
[]
``` 

and that the sub app automatically understands to get the folder with '*' corresponding to the instantiation index of the sub-app. Or something to that effect.

## Impact

This feature seems essential to many multiphysics problems with transients restarting from steady-states.

### PR DESCRIPTION
**Pull request by vincentlaboure**
This MR:

- fixes previous tests that used checkpoints at both the master and sub-app level
- adds a steady to transient restart test with multiple sub apps
- adds a pseudo-transient to transient restart test with multiple sub apps

Close #15287

### CODE REVIEW

**Comment by vincentlaboure** [comment]
This is now ready for review.

**Comment by vincentlaboure** [comment]
Would anyone be willing to review this? @permcody or @fdkong ?

**Comment by permcody** [comment]
Thanks @vincentlaboure. One of us will get to this. My morning is packed.

File: test/tests/restart/restart_transient_from_transient/restart_trans_with_2subs_sub.i [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+[Executioner]
+  type = Transient
+
+  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart '
+  petsc_options_value = 'hypre boomeramg 100'
+
+  start_time = 0
+  end_time = 3
+  dt = 1.0
+  # steady_state_detection = true
```
Comment (loganharbour) [inline test/tests/restart/restart_transient_from_transient/restart_trans_with_2subs_sub.i:60]:
```suggestion
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Sorry that this sat around for so long. Behavior looks a lot better than before.

Please just remove the old steady state flag and this should be good to go.

**Review by fdkong — COMMENTED** [review]
Looks great for me

**Review by loganharbour — APPROVED** [review]

**Comment by vincentlaboure** [comment]
Thanks @loganharbour !

**Pull request merged**